desc/compute/_stability.py

```diff
@@ -231,3 +231,513 @@ def _magnetic_well(params, transforms, profiles, data, **kwargs):
+
+    gds2 = jnp.reshape(
+        rho**2
+        * (
+            data["g^aa"][None, :]
+            - 2 * sign_iota * shat / rho * zeta0[:, None] * data["g^ra"][None, :]
+            + zeta0[:, None] ** 2 * (shat / rho) ** 2 * data["g^rr"][None, :]
+        ),
+        (N_alpha, N_zeta0, N_zeta),
+    )
+
+    g = a_N**3 * B_N * gds2 / B * gradpar
+    g_half = (g[:, :, 1:] + g[:, :, :-1]) / 2
+    c = (
+        1
+        * a_N**3
+        * B_N
+        * jnp.reshape(
+            2
+            / data["B^zeta"][None, :]
+            * sign_psi
+            * rho**2
+            * dpdpsi
+            * (
+                data["cvdrift"][None, :]
+                - shat / rho * zeta0[:, None] * data["cvdrift0"][None, :]
+            ),
+            (N_alpha, N_zeta0, N_zeta),
+        )
+    )
+
+    h = phi[1] - phi[0]
+
+    # g_half on half grid points, c_full on full grid points
+    g_half = (g[:, :, 1:] + g[:, :, :-1]) / 2
+    c_full = c[:, :, :-1]
+
+    i = jnp.arange(N_alpha)[:, None, None]
+    j = jnp.arange(N_zeta0)[None, :, None]
+    k = jnp.arange(N_zeta - 1)[None, None, :]
+
+    X = jnp.zeros((N_alpha, N_zeta0, N_zeta - 1))
+    X = X.at[i, j, k].set(phi[k])
+
+    Y = jnp.zeros((N_alpha, N_zeta0))
+    eps = 5e-3  # slope of the test function
+    Yp = eps * jnp.ones((N_alpha, N_zeta0))
+
+    @jit
+    def integrator(carry, x):
+        arr, arr_d = carry
+        g_element, c_element = x
+        # Update the array (Y) and its derivative
+        arr_updated = (g_element * arr + arr_d * h) / g_element
+        arr_d_updated = arr_d - c_element * arr_updated * h
+        # Calculate the sign of the product of arr and arr_updated
+        sign_product = jnp.sign(arr * arr_updated)
+
+        return (arr_updated, arr_d_updated), (arr_updated, sign_product)
+
+    @jit
+    def cumulative_update_jit(arr, arr_d, g_half, c_full):
+        _, scan_output = scan(integrator, (arr, arr_d), (g_half, c_full))
+        Y, sign_product = scan_output
+        # Create a mask where sign_product is negative
+        negative_mask = sign_product < 0
+        # Find the first occurrence of a negative value
+        first_negative_index = jnp.argmax(negative_mask)
+        # Use where to return 0 if there are no negative values
+        first_negative_index = jnp.where(
+            jnp.any(negative_mask), first_negative_index, -1
+        )
+        # This factor will give us the exact X point of intersection
+        lin_interp_factor = jnp.where(
+            first_negative_index != -1,
+            1 / (1 - Y[first_negative_index + 1] / Y[first_negative_index]),
+            0,
+        )
+
+        return Y, first_negative_index, lin_interp_factor
+
+    # Vectorize over the first two dimensions
+    vectorized_cumulative_update = jit(vmap(vmap(cumulative_update_jit)))
+    Y, first_negative_indices, lin_interp_factors = vectorized_cumulative_update(
+        Y, Yp, g_half, c_full
+    )
```

Comment:
It'd be better to define these functions as private functions outside the compute function, decorated with jnp.vectorize rather than a double vmap and double jit. Then you can jit the compute function itself which will be faster.